tracer/src/Datadog.Trace/Telemetry/Collectors/DependencyTelemetryCollector.cs

```diff
@@ -71,12 +69,23 @@ public bool HasChanges()
         public ICollection<DependencyTelemetryData> GetData()
         {
             var hasChanges = Interlocked.CompareExchange(ref _hasChangesFlag, 0, 1) == 1;
+
             if (!hasChanges)
             {
                 return null;
             }
 
-            return _assemblies.Keys;
+            var assembliesToRecord = new List<DependencyTelemetryData>();
```

Comment:
[heap allocations] Is this `GetData()` method called often? How hard would it be to track the count of new assemblies so you could initialize the list with the correct count? Or even use an array instead of a `List<T>`?

```csharp
var assembliesToRecord = new List<DependencyTelemetryData>(newAssemblyCount);

var assembliesToRecord = new DependencyTelemetryData[newAssemblyCount];
```

Reply:
> Is this GetData() method called often?

Every 1 minute (max)

> How hard would it be to track the count of new assemblies

_Moderately_ hard I think with the current design due to concurrency. The trouble is that assemblies could be added to the underlying dictionary while the following loop is created. 

Alternative suggestion - I believe (please double check) calls to this method are single-threaded. So you could do something like this:

```csharp
private List<DependencyTelemetryData>? _toSend;

public ICollection<DependencyTelemetryData> GetData()
{
    // ...
    _toSend ??= new List<DependencyTelemetryData>(_assemblies.Count);
    _toSend.Clear();

    foreach (var assembly in _assemblies)
    //...
```

That way you initialize to the initial set, grow as necessary, but don't create a new list each time